Netcracker/KubeMarine
# [CPDEV-95324] Update 3rd-parties

### Description
* Support Kubernetes versions
    * v1.26.11
    * v1.27.8
    * v1.28.4
* Increase patch versions for all plugins mapped to Kubernetes v1.26.11 - v1.28.3
* Update crictl to v1.28.0 mapped to Kubernetes v1.26.11 - v1.28.3
* Support nginx-ingress-controller v1.9.4 for Kubernetes v1.28.4
    * Set `allow-snippet-annotations: "true"` for backward compatibility.
* Support python 3.12
    * Update docker base image to `python:3.12-slim-bullseye`
* Increased versions of 3rd-party packages.
    * major versions of fabric, paramiko, invoke.
* Helm in docker image is increased to v3.13.2
* Updated dependency on cryptography==41.0.7
    * Fixed building of executable for macOS arm64 platform.

### How to apply
Run `kubemarine migrate_kubemarine --force-apply pin_kubernetes_version,upgrade_crictl,upgrade_calico,upgrade_nginx_ingress_controller,upgrade_local_path_provisioner`.

### Release notes
Kubemarine offers the following configuration to *partially* mitigate ingress-nginx vulnerabilities.
```yaml
plugins:
  nginx-ingress-controller:
    config_map:
      # CVE-2021-25742
      allow-snippet-annotations: "false"
      # CVE-2022-4886
      # Available in ingress-nginx supported for Kubernetes v1.26.7 or higher
      strict-validate-path-type: "true"
    controller:
      args:
        # CVE-2023-5043, CVE-2023-5044
        # Available in ingress-nginx supported for Kubernetes v1.28.4 or higher
        - --enable-annotation-validation  # CVE-2023-5043, CVE-2023-5044
```

**Note:** Kubemarine does not support `kubernetes-dashboard` plugin if you disable the `allow-snippet-annotations` property.

For more information, refer to
[CVE-2021-25742](https://github.com/advisories/GHSA-4pp2-3663-mcw8)
[CVE-2022-4886](https://github.com/advisories/GHSA-gvrm-w2f9-f77q)
[CVE-2023-5043](https://github.com/advisories/GHSA-5wj4-wffq-3378)
[CVE-2023-5044](https://github.com/advisories/GHSA-fp9f-44c2-cw27)

### Test Cases

**TestCase 1**

Install or upgrade to Kubernetes v1.26.11, v1.27.8, v1.28.4

**TestCase 2**

Upgrade plugins and crictl on existing clusters.

Steps:

1. Install v1.27.1 or higher using **previous** version of Kubemarine.
2. Run `check_paas` using **new** version of Kubemarine.

ER: Some check fail because plugin versions are outdated.

3. Run `kubemarine migrate_kubemarine --force-apply upgrade_crictl,upgrade_calico,upgrade_nginx_ingress_controller,upgrade_local_path_provisioner`

ER: Plugins and crictl are upgraded.

4. Run `check_paas`

ER: All checks are successful.

**TestCase 3**

Set previous default in the inventory.

Steps:

1. Install using **previous** version of Kubemarine not specifying `services.kubeadm.kubernetesVersion` in the inventory.

ER: Kubernetes v1.26.7 is installed.

2. Run `kubemarine migrate_kubemarine --force-apply pin_kubernetes_version` using **new** Kubemarine.

ER: `services.kubeadm.kubernetesVersion` is set to v1.26.11 in the inventory.

### Checklist

#### Unit tests
test/unit/plugins/ - update tests to support new plugin versions.
